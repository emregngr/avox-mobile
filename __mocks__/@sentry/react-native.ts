import type { ComponentType } from 'react'

const mockedSentry = {
  init: jest.fn(),
  wrap: <T extends ComponentType<any>>(Component: T): T => Component,
  withScope: jest.fn((callback: (scope: any) => void) =>
    callback({
      setTag: jest.fn(),
      setLevel: jest.fn(),
      setUser: jest.fn(),
      setContext: jest.fn(),
    }),
  ),
  captureException: jest.fn(),
  captureMessage: jest.fn(),
  addBreadcrumb: jest.fn(),
  configureScope: jest.fn(),
  setUser: jest.fn(),
  setTag: jest.fn(),
  setContext: jest.fn(),
  setRelease: jest.fn(),
  setDist: jest.fn(),
  setEnvironment: jest.fn(),
  close: jest.fn(() => Promise.resolve(true)),
  flush: jest.fn(() => Promise.resolve(true)),
}

module.exports = mockedSentry

export type mockedSentryType = typeof mockedSentry
