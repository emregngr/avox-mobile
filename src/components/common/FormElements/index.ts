export * from './CheckboxField'
export * from './TextInputField'
