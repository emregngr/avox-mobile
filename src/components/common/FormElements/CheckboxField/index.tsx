import { Ionicons } from '@expo/vector-icons'
import React, { useCallback, useMemo } from 'react'
import type { Control } from 'react-hook-form'
import { Controller } from 'react-hook-form'
import { Pressable, TouchableOpacity, View } from 'react-native'

import { ThemedText } from '@/components/common/ThemedText'
import { getLocale } from '@/locales/i18next'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'

interface CheckboxFieldProps {
  control: Control<any>
  disabled?: boolean
  error?: string
  labelKey: string
  name: string
  onPressLabel?: () => void
}

export const CheckboxField = ({
  control,
  disabled = false,
  error,
  labelKey,
  name,
  onPressLabel,
}: CheckboxFieldProps) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const getCheckboxColor = useCallback(
    (error: any, value: boolean) => {
      if (error) return colors?.error
      if (value) return colors?.success
      return colors?.onPrimary100
    },
    [colors],
  )

  return (
    <View>
      <View className="flex-row items-center">
        <Controller
          render={({ field: { onChange, value } }) => (
            <TouchableOpacity
              activeOpacity={0.7}
              className="p-1"
              disabled={disabled}
              hitSlop={10}
              onPress={() => onChange(!value)}
            >
              <View className="relative">
                <Ionicons color={getCheckboxColor(error, value)} name="square-outline" size={24} />
                {value ? (
                  <Ionicons
                    className="absolute top-0.5 left-0.5"
                    color={colors?.success}
                    name="checkmark-sharp"
                    size={20}
                  />
                ) : null}
              </View>
            </TouchableOpacity>
          )}
          control={control}
          name={name}
        />

        <Pressable className="flex-1 ml-3" hitSlop={10} onPress={onPressLabel}>
          <ThemedText className="underline" color="text-100" type="body1">
            {getLocale(labelKey)}
          </ThemedText>
        </Pressable>
      </View>

      {error ? (
        <ThemedText className="mt-1 ml-12" color="error" type="body2">
          {error}
        </ThemedText>
      ) : null}
    </View>
  )
}
