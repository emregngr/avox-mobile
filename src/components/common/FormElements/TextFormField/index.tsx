import { Ionicons } from '@expo/vector-icons'
import type { ReactElement, Ref } from 'react'
import React, { forwardRef, useMemo, useState } from 'react'
import type { Control, FieldValues, Path } from 'react-hook-form'
import { Controller } from 'react-hook-form'
import type { KeyboardTypeOptions } from 'react-native'
import { TextInput, TouchableOpacity, View } from 'react-native'

import ColoredClear from '@/assets/icons/coloredClear'
import { ThemedText } from '@/components/common/ThemedText'
import useThemeStore from '@/store/theme'
import { themeColors } from '@/themes'
import { cn } from '@/utils/common/cn'

type AutoCapitalizeType = 'none' | 'sentences' | 'words' | 'characters'
type ReturnKeyType = 'done' | 'go' | 'next' | 'search' | 'send'
type TextContentType =
  | 'none'
  | 'emailAddress'
  | 'password'
  | 'newPassword'
  | 'givenName'
  | 'familyName'
  | 'name'
  | 'username'
type AutoCompleteType =
  | 'off'
  | 'email'
  | 'password'
  | 'given-name'
  | 'family-name'
  | 'name'
  | 'username'
type SubmitBehavior = 'submit' | 'blurAndSubmit' | 'newline'

interface TextFormFieldProps<T extends FieldValues> {
  autoCapitalize?: AutoCapitalizeType
  autoCorrect?: boolean
  clearTextOnFocus?: boolean
  control: Control<T>
  editable?: boolean
  keyboardType?: KeyboardTypeOptions
  label: string
  name: Path<T>
  onSubmitEditing?: () => void
  placeholder: string
  returnKeyType?: ReturnKeyType
  secureTextEntry?: boolean
  showToggle?: boolean
  submitBehavior?: SubmitBehavior
  textContentType?: TextContentType
}

const TextFormFieldComponent = <T extends FieldValues>(
  {
    autoCapitalize = 'none',
    autoCorrect = false,
    clearTextOnFocus = false,
    control,
    editable = true,
    keyboardType = 'default',
    label,
    name,
    onSubmitEditing,
    placeholder,
    returnKeyType = 'next',
    secureTextEntry = false,
    showToggle = false,
    submitBehavior = 'blurAndSubmit',
    textContentType,
  }: TextFormFieldProps<T>,
  ref: Ref<TextInput>,
) => {
  const { selectedTheme } = useThemeStore()

  const colors = useMemo(() => themeColors?.[selectedTheme], [selectedTheme])

  const [isPasswordVisible, setIsPasswordVisible] = useState<boolean>(false)
  const [isFocused, setIsFocused] = useState<boolean>(false)

  const getAutoCompleteType = (): AutoCompleteType => {
    const fieldName = name.toLowerCase()
    if (fieldName.includes('email')) return 'email'
    if (fieldName.includes('password')) return 'password'
    if (fieldName.includes('firstname') || fieldName.includes('givenname')) return 'given-name'
    if (fieldName.includes('lastname') || fieldName.includes('familyname')) return 'family-name'
    if (fieldName.includes('name')) return 'name'
    if (fieldName.includes('username')) return 'username'
    return 'off'
  }

  const getTextContentType = (): TextContentType => {
    if (textContentType) return textContentType

    const fieldName = name.toLowerCase()
    if (fieldName.includes('email')) return 'emailAddress'
    if (fieldName.includes('password')) return secureTextEntry ? 'password' : 'none'
    if (fieldName.includes('firstname') || fieldName.includes('givenname')) return 'givenName'
    if (fieldName.includes('lastname') || fieldName.includes('familyname')) return 'familyName'
    if (fieldName.includes('name')) return 'name'
    if (fieldName.includes('username')) return 'username'
    return 'none'
  }

  return (
    <Controller
      render={({ field: { onBlur, onChange, value }, fieldState: { error } }) => (
        <View className="w-full">
          <ThemedText className="my-1.5 px-1" color="text-70" type="body1">
            {label}
          </ThemedText>
          <View
            className={cn(
              'flex-row items-center justify-center px-4 rounded-xl overflow-hidden bg-background-tertiary border',
              error ? 'border-error' : 'border-transparent',
            )}
          >
            <TextInput
              className={cn(
                'flex-1 py-3 text-text-100 placeholder:text-text-50 text-[16px] font-inter-medium',
              )}
              onBlur={() => {
                setIsFocused(false)
                onBlur()
              }}
              passwordRules={
                secureTextEntry
                  ? 'minlength: 6; required: lower; required: upper; required: digit;'
                  : undefined
              }
              allowFontScaling={false}
              autoCapitalize={autoCapitalize}
              autoComplete={getAutoCompleteType()}
              autoCorrect={autoCorrect}
              clearTextOnFocus={clearTextOnFocus}
              editable={editable}
              keyboardAppearance={selectedTheme === 'dark' ? 'dark' : 'light'}
              keyboardType={keyboardType}
              maxFontSizeMultiplier={1.0}
              onChangeText={onChange}
              onFocus={() => setIsFocused(true)}
              onSubmitEditing={onSubmitEditing}
              placeholder={placeholder}
              ref={ref}
              returnKeyType={returnKeyType}
              secureTextEntry={secureTextEntry && !isPasswordVisible}
              spellCheck={false}
              submitBehavior={submitBehavior}
              textAlignVertical="center"
              textContentType={getTextContentType()}
              underlineColorAndroid="transparent"
              value={value}
              enablesReturnKeyAutomatically
            />
            {secureTextEntry && showToggle ? (
              <TouchableOpacity
                activeOpacity={0.7}
                className={isFocused && value?.length > 0 ? 'mr-3' : ''}
                hitSlop={10}
                onPress={() => setIsPasswordVisible(prev => !prev)}
              >
                <Ionicons
                  color={colors?.onPrimary100}
                  name={isPasswordVisible ? 'eye-off' : 'eye'}
                  size={20}
                />
              </TouchableOpacity>
            ) : null}
            {isFocused && value?.length > 0 ? (
              <TouchableOpacity activeOpacity={0.7} hitSlop={10} onPress={() => onChange('')}>
                <ColoredClear
                  height={20}
                  primaryColor={colors?.background?.primary}
                  secondaryColor={colors?.onPrimary100}
                  width={20}
                />
              </TouchableOpacity>
            ) : null}
          </View>
          {error ? (
            <ThemedText className="my-1.5 px-1" color="error" type="body2">
              {error?.message}
            </ThemedText>
          ) : null}
        </View>
      )}
      control={control}
      name={name}
    />
  )
}

export const TextFormField = forwardRef(TextFormFieldComponent) as <T extends FieldValues>(
  props: TextFormFieldProps<T> & { ref?: Ref<TextInput> },
) => ReactElement
