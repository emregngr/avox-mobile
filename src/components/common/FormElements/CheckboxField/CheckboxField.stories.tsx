import type { Meta, StoryObj } from '@storybook/react-native'
import { useForm } from 'react-hook-form'
import { Alert, View } from 'react-native'

import { CheckboxField } from '@/components/common/FormElements/CheckboxField'

interface FormData {
  termsAccepted: boolean
  marketingEmails: boolean
  dataProcessing: boolean
  newsletter: boolean
}

const meta: Meta<typeof CheckboxField> = {
  title: 'Components/Forms/CheckboxField',
  component: CheckboxField,
  parameters: {
    docs: {
      description: {
        component: 'Checkbox field component with form validation support using react-hook-form.',
      },
    },
  },
  decorators: [
    Story => (
      <View className="p-4 bg-background-primary flex-1">
        <Story />
      </View>
    ),
  ],
  argTypes: {
    disabled: {
      control: { type: 'boolean' },
      description: 'Disable the checkbox interaction',
    },
    error: {
      control: { type: 'text' },
      description: 'Error message to display',
    },
    labelKey: {
      control: { type: 'text' },
      description: 'Translation key for the label',
    },
    name: {
      control: { type: 'text' },
      description: 'Form field name',
    },
    testID: {
      control: { type: 'text' },
      description: 'Test identifier',
    },
  },
}

export default meta

type Story = StoryObj<typeof CheckboxField>

const FormWrapper = ({
  name = 'termsAccepted',
  labelKey = 'I agree to the terms and conditions',
  error,
  disabled = false,
  onPressLabel,
  testID,
}: {
  name?: keyof FormData
  labelKey?: string
  error?: string
  disabled?: boolean
  onPressLabel?: () => void
  testID?: string
}) => {
  const { control } = useForm<FormData>({
    defaultValues: {
      termsAccepted: false,
      marketingEmails: true,
      dataProcessing: false,
      newsletter: false,
    },
  })

  return (
    <CheckboxField
      control={control}
      name={name}
      labelKey={labelKey}
      disabled={disabled}
      {...(error && { error })}
      {...(onPressLabel && { onPressLabel })}
      {...(testID && { testID })}
    />
  )
}

export const Default: Story = {
  render: () => <FormWrapper />,
  parameters: {
    docs: {
      description: {
        story: 'Basic checkbox field with form integration.',
      },
    },
  },
}

export const Checked: Story = {
  render: () => <FormWrapper name="marketingEmails" labelKey="Subscribe to marketing emails" />,
  parameters: {
    docs: {
      description: {
        story: 'Checkbox field in checked state (default value is true).',
      },
    },
  },
}

export const WithError: Story = {
  render: () => (
    <FormWrapper
      labelKey="I agree to the privacy policy"
      error="You must accept the privacy policy to continue"
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Checkbox field with validation error message.',
      },
    },
  },
}

export const Disabled: Story = {
  render: () => <FormWrapper labelKey="This option is currently unavailable" disabled={true} />,
  parameters: {
    docs: {
      description: {
        story: 'Disabled checkbox field that cannot be interacted with.',
      },
    },
  },
}

export const LongLabel: Story = {
  render: () => (
    <FormWrapper labelKey="I agree to the very long terms and conditions that may span multiple lines and test the text wrapping behavior of the component" />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Checkbox field with long label text to test wrapping.',
      },
    },
  },
}

export const ClickableLabel: Story = {
  render: () => (
    <FormWrapper
      labelKey="Click this label to open terms"
      onPressLabel={() => Alert.alert('Label Clicked', 'Terms and conditions would open here')}
    />
  ),
  parameters: {
    docs: {
      description: {
        story: 'Checkbox field with clickable label that triggers an action.',
      },
    },
  },
}

export const MultipleCheckboxes: Story = {
  render: () => {
    const { control } = useForm<FormData>({
      defaultValues: {
        termsAccepted: false,
        marketingEmails: true,
        dataProcessing: false,
        newsletter: false,
      },
    })

    const checkboxes = [
      {
        name: 'termsAccepted' as keyof FormData,
        labelKey: 'I agree to the terms and conditions',
      },
      {
        name: 'marketingEmails' as keyof FormData,
        labelKey: 'Send me marketing emails',
      },
      {
        name: 'dataProcessing' as keyof FormData,
        labelKey: 'Allow data processing for analytics',
        error: 'This field is required for service functionality',
      },
      {
        name: 'newsletter' as keyof FormData,
        labelKey: 'Subscribe to weekly newsletter',
      },
    ]

    return (
      <View className="gap-4">
        {checkboxes.map((checkbox, index) => {
          const onPressLabel =
            checkbox.name === 'termsAccepted'
              ? () => Alert.alert('Terms', 'Would open terms and conditions')
              : undefined

          return (
            <CheckboxField
              key={checkbox.name}
              control={control}
              name={checkbox.name}
              labelKey={checkbox.labelKey}
              testID={`checkbox-${index}`}
              {...(checkbox.error && { error: checkbox.error })}
              {...(onPressLabel && { onPressLabel })}
            />
          )
        })}
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'Multiple checkbox fields in a form with various states and errors.',
      },
    },
  },
}

export const AllStates: Story = {
  render: () => {
    const { control } = useForm({
      defaultValues: {
        normal: false,
        checked: true,
        error: false,
        disabled: false,
      },
    })

    return (
      <View className="gap-4">
        <View>
          <CheckboxField control={control} name="normal" labelKey="Normal unchecked state" />
        </View>

        <View>
          <CheckboxField control={control} name="checked" labelKey="Checked state" />
        </View>

        <View>
          <CheckboxField
            control={control}
            name="error"
            labelKey="With error state"
            error="This field has an error"
          />
        </View>

        <View>
          <CheckboxField
            control={control}
            name="disabled"
            labelKey="Disabled state"
            disabled={true}
          />
        </View>
      </View>
    )
  },
  parameters: {
    docs: {
      description: {
        story: 'All possible states of the checkbox field in one view.',
      },
    },
  },
}

export const Playground: Story = {
  render: (args: any) => {
    const { name, labelKey, disabled, testID, error, onPressLabel } = args
    return (
      <FormWrapper
        name={name as keyof FormData}
        labelKey={labelKey}
        disabled={disabled}
        {...(error && { error })}
        {...(onPressLabel && { onPressLabel })}
        {...(testID && { testID })}
      />
    )
  },
  args: {
    name: 'termsAccepted',
    labelKey: 'I agree to the terms and conditions',
    disabled: false,
    error: '',
    testID: 'playground-checkbox',
  },
  parameters: {
    docs: {
      description: {
        story: 'Interactive playground to test all props and configurations.',
      },
    },
  },
}
