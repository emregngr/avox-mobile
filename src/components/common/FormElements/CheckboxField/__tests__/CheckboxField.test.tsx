import { fireEvent, render } from '@testing-library/react-native'
import React from 'react'
import { useForm } from 'react-hook-form'

import { CheckboxField } from '@/components/common/FormElements/CheckboxField'
import useThemeStore from '@/store/theme'

jest.mock('@/locales/i18next', () => ({
  getLocale: (key: string) => key,
}))

jest.mock('@/store/theme')

const mockedUseThemeStore = useThemeStore as jest.MockedFunction<typeof useThemeStore>

const MockedCheckboxField = (props: any) => {
  const { control } = useForm({
    defaultValues: { test: false },
  })

  return <CheckboxField control={control} name="test" {...props} />
}

beforeEach(() => {
  mockedUseThemeStore.mockReturnValue({
    selectedTheme: 'light',
  })
})

describe('CheckboxField Component', () => {
  it('renders unchecked checkbox initially', () => {
    const { getByTestId } = render(<MockedCheckboxField labelKey="label.test" />)
    expect(getByTestId('checkbox')).toBeTruthy()
  })

  it('toggles checkbox when pressed', () => {
    const { getByTestId } = render(<MockedCheckboxField labelKey="label.test" />)
    const checkbox = getByTestId('checkbox')

    fireEvent.press(checkbox)

    expect(checkbox).toBeTruthy()
  })

  it('calls onPressLabel when label is pressed', () => {
    const mockedFn = jest.fn()
    const { getByText } = render(
      <MockedCheckboxField labelKey="label.test" onPressLabel={mockedFn} />,
    )

    fireEvent.press(getByText('label.test'))
    expect(mockedFn).toHaveBeenCalled()
  })

  it('shows error message when error is provided', () => {
    const { getByText } = render(
      <MockedCheckboxField error="This is error" labelKey="label.test" />,
    )
    expect(getByText('This is error')).toBeTruthy()
  })
})

describe('CheckboxField Component Snapshot', () => {
  it('should render the CheckboxField Component successfully', () => {
    const { toJSON } = render(<MockedCheckboxField labelKey="label.test" />)

    expect(toJSON()).toMatchSnapshot()
  })
})
