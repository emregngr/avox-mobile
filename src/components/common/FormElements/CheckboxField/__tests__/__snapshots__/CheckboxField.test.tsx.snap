// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CheckboxField Component Snapshot should render the CheckboxField Component successfully 1`] = `
<View>
  <View
    className="flex-row items-center"
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": false,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={10}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="checkbox"
    >
      <View
        color="#000000"
        name="checkbox-blank-outline"
        size={24}
        testID="mocked-material-community-icon"
      />
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      className="flex-row items-center flex-1 ml-3"
      collapsable={false}
      focusable={true}
      hitSlop={10}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <Text
        allowFontScaling={false}
        className="text-[16px] leading-[19px] font-inter-medium text-text-100 underline"
        maxFontSizeMultiplier={1}
      >
        label.test
      </Text>
    </View>
  </View>
</View>
`;
